### PR TITLE
Retry forever for Seed and RebuildCache

### PR DESCRIPTION
**Pull request by v-limliu**
# Type of change? 

- [x] Bug fix
- [ ] New feature
- [ ] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 

While being deployed with Kubernates (without the dependsOn feature which supported only in docker), the Seed feature will failed because it could not resolve the underlying service host name at the beginning. Here we are going to add logic to try it forever, regardless any exception raised, to ensure the Seed and RebuildCache actions will be invoked.

**Checklist:**

- [x] All tests passed
- [ ] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java [hunk 1]
```diff
@@ -37,8 +37,10 @@
     private static final Logger.ALogger log = Logger.of(Cache.class);
     private final int cacheTTL;
     private final int rebuildTimeout;
+    private final int rebuildCacheRetryInterval = 10;
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java]: could you make it explicit whether it's seconds or minutes?

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java [hunk 2]
```diff
@@ -37,8 +37,10 @@
...
     private final String CacheCollectionId = "cache";
     private final String CacheKey = "twin";
+    private final int rebuildRetry = 5;
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java]: retry only 5 times? shouldn't it retry forever?

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java [hunk 3]
```diff
@@ -158,10 +152,9 @@ public CompletionStage RebuildCacheAsync(boolean force) throws BaseException, Ex
                     reportedNames = reportedNamesTask.get();
                     reportedNames.addAll(twinNames.getReportedProperties());
                 } catch (Exception e) {
-                    log.info(String.format("retry %d for %s:%s  IothubManagerService and SimulationService  are not both ready,wait 10 seconds", retry, CacheCollectionId, CacheKey));
+                    log.info(String.format("retry %d for %s:%s  IothubManagerService and SimulationService  are not both ready,wait 10 seconds", retry, CacheCollectionId, CacheKey), e);
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java]: log as a warning
nit: add spaces after ","
nit: split long lines

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java [hunk 4]
```diff
@@ -158,10 +152,9 @@ public CompletionStage RebuildCacheAsync(boolean force) throws BaseException, Ex
...
                     if (retry-- < 1) {
-                        return CompletableFuture.runAsync(() -> {
-                        });
+                        throw new BaseException(String.format("failed to rebuild cache after %s retries", rebuildRetry));
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java]: let's not use BaseException please, it doesn't really help handling errors. Exceptions should be very specific

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java [hunk 5]
```diff
@@ -35,6 +35,7 @@
     private final String MutexKey = "seedMutex";
     private final String CompletedFlagKey = "seedCompleted";
     private int mutexTimeout = 60 * 5;
+    private final int seedRetryInterval = 10;
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: secs/minutes?

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java [hunk 6]
```diff
@@ -35,6 +35,7 @@
     private final String MutexKey = "seedMutex";
     private final String CompletedFlagKey = "seedCompleted";
     private int mutexTimeout = 60 * 5;
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: seconds/minutes?

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java [hunk 7]
```diff
@@ -101,12 +94,12 @@ public CompletionStage TrySeedAsync() throws ExternalDependencyException {
             this.log.info("Seed end");
             this.SetCompletedFlagAsync().toCompletableFuture().get();
             this.mutex.LeaveAsync(SeedCollectionId, MutexKey).toCompletableFuture().get();
-        } catch (InterruptedException | ExecutionException | BaseException e) {
+            return CompletableFuture.runAsync(() -> {
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: quite sure there is a better way to return "nothing" :-)

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java [hunk 8]
```diff
@@ -101,12 +94,12 @@ public CompletionStage TrySeedAsync() throws ExternalDependencyException {
...
+            });
+        } catch (Exception e) {
             log.error("Seed failed");
             throw new ExternalDependencyException("Seed failed");
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: include the original exception

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java [hunk 9]
```diff
@@ -101,12 +94,12 @@ public CompletionStage TrySeedAsync() throws ExternalDependencyException {
             this.log.info("Seed end");
             this.SetCompletedFlagAsync().toCompletableFuture().get();
             this.mutex.LeaveAsync(SeedCollectionId, MutexKey).toCompletableFuture().get();
-        } catch (InterruptedException | ExecutionException | BaseException e) {
+            return CompletableFuture.runAsync(() -> {
+            });
+        } catch (Exception e) {
             log.error("Seed failed");
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: log the exception

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java [hunk 10]
```diff
@@ -357,15 +350,19 @@ private CompletionStage SeedSingleTemplateAsync(String content) throws InvalidIn
             try {
                 this.storage.updateDeviceGroupAsync(m.getId(), m, "*");
             } catch (Exception ex) {
-                this.log.error(String.format("Failed to seed default group {group.DisplayName}", m.getDisplayName()));
+                String errorMessage = String.format("Failed to seed default group %s", m.getDisplayName());
+                this.log.error(errorMessage);
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: log the exception

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java [hunk 11]
```diff
@@ -357,15 +350,19 @@ private CompletionStage SeedSingleTemplateAsync(String content) throws InvalidIn
...
+                throw new RuntimeException(errorMessage);
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: use a more specific exception and include the original exception

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java [hunk 12]
```diff
@@ -357,15 +350,19 @@ private CompletionStage SeedSingleTemplateAsync(String content) throws InvalidIn
...
             }
         });
 
         StreamSupport.stream(template.getRules().spliterator(), false).forEach(m -> {
             try {
                 this.telemetryClient.UpdateRuleAsync(m, "*");
             } catch (Exception ex) {
-                this.log.error(String.format("Failed to seed default rule %s", m.getDescription()));
+                String errorMessage = String.format("Failed to seed default rule %s", m.getDescription());
+                this.log.error(errorMessage);
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: log the exception

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java [hunk 13]
```diff
@@ -357,15 +350,19 @@ private CompletionStage SeedSingleTemplateAsync(String content) throws InvalidIn
...
+                throw new RuntimeException(errorMessage);
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: use a more specific exception and include the original exception

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java [hunk 14]
```diff
@@ -379,7 +376,9 @@ private CompletionStage SeedSingleTemplateAsync(String content) throws InvalidIn
                 this.simulationClient.UpdateSimulationAsync(simulationModel).toCompletableFuture().get();
             }
         } catch (Exception ex) {
-            this.log.info("Failed to seed default simulation");
+            String errorMessage = "Failed to seed default simulation";
+            this.log.info(errorMessage);
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: log as error

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java [hunk 15]
```diff
@@ -379,7 +376,9 @@ private CompletionStage SeedSingleTemplateAsync(String content) throws InvalidIn
...
+            throw new RuntimeException(errorMessage);
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: use a more specific exception and include the original exception

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/external/ValueApiModel.java [hunk 16]
```diff
@@ -41,7 +41,7 @@ public void setData(String data) {
         this.data = data;
     }
 
-    @JsonProperty("ETag")
+    @JsonProperty("Etag")
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/external/ValueApiModel.java]: this is incorrect. "ETag" is the correct spelling. Plus this change could break clients

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/helpers/RetryHelper.java [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+    private static final Logger.ALogger log = Logger.of(RetryHelper.class);
+
+    public static void retry(String name, Object entry, long interval, long timeout) {
+        DateTime startTime = DateTime.now();
+        while (Seconds.secondsBetween(startTime, DateTime.now()).getSeconds() < timeout) {
+            try {
+                if (entry instanceof ICache) {
+                    ((ICache) entry).RebuildCacheAsync().toCompletableFuture().get();
+                }
+                if (entry instanceof ISeed) {
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/helpers/RetryHelper.java]: nit: use `else`

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/helpers/RetryHelper.java [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+                    ((ISeed) entry).TrySeedAsync().toCompletableFuture().get();
+                }
+                return;
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/helpers/RetryHelper.java]: if entry is neither of ICache or ISeed, there should be an exception.

Btw, I was expecting a generic helper, but the class is tightly coupled to ICache and ISeed, which make the overall design hard to follow

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java (hunk 1 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java (hunk 2 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java]: 5 times is a former retry logic, now it will be an unit for forever retry .  For clear , I remove this code here.

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java (hunk 3 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java (hunk 4 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java (hunk 5 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java (hunk 6 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java (hunk 7 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java (hunk 8 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java (hunk 10 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/helpers/RetryHelper.java (hunk 18 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/helpers/RetryHelper.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/helpers/RetryHelper.java (hunk 17 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/helpers/RetryHelper.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java (hunk 11 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: Unchecked exceptions are  better for stream API

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java (hunk 13 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: Unchecked exceptions are  better for stream API

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java (hunk 15 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: Unchecked exceptions are  better for stream API

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java (hunk 12 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java (hunk 14 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/external/ValueApiModel.java (hunk 16 above)
Comment (zhang-hua) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/external/ValueApiModel.java]: this is a client code to access storage adapter service which is using 'Etag'. it must be aligned with storage adapter: https://github.com/Azure/pcs-storage-adapter-java/blob/master/app/com/microsoft/azure/iotsolutions/storageadapter/webservice/v1/models/ValueApiModel.java#L46

**Comment by zhang-hua** [comment]
@dluc , we have confirmed the exception handle and retry logic meet the requirements as we discussed in the email thread. Any exception thrown by Seed class will be captured by RecurringTask and seed process will be retried for ever it any error happened. The Dockerized container has also been tested to verify the fix.

File: app/Module.java [hunk 19]
```diff
@@ -27,8 +24,7 @@ public void configure() {
         // Note: this method should be empty
         // Try to use use JIT binding and @ImplementedBy instead
         // global setting is not recommend for application_onStart event, PLS refer here for details :https://www.playframework.com/documentation/2.6.x/GlobalSettings
-        bind(ICache.class).to(Cache.class).asEagerSingleton();
-        bind(ISeed.class).to(Seed.class).asEagerSingleton();
+        bind(IRecurringTasks.class).to(RecurringTasks.class).asEagerSingleton();
```
Comment (dluc) [inline app/Module.java]: could you explain why we need the eager creation?

File: app/Module.java [hunk 20]
```diff
@@ -2,10 +2,7 @@
```
Comment (dluc) [inline app/Module.java]: code style: java methods should start with a lowercase letter

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java [hunk 21]
```diff
@@ -129,66 +116,55 @@ public void run() {
...
     @Override
-    public CompletionStage RebuildCacheAsync(boolean force) throws BaseException, ExecutionException, InterruptedException, URISyntaxException {
-        {
-            int retry = 5;
-            while (true) {
-                HashSet<String> reportedNames = null;
-                DeviceTwinName twinNames = null;
-                ValueApiModel cache = null;
-                String etag = null;
+    public CompletionStage RebuildCacheAsync(boolean force) throws Exception {
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java]: we should specify explicitly all the checked exceptions, otherwise it becomes impossible to handle them appropriately upstream

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java [hunk 22]
```diff
@@ -129,66 +116,55 @@ public void run() {
...
-                        return CompletableFuture.runAsync(() -> {
-                        });
-                    }
-                    Thread.sleep(10000);
-                    continue;
-                }
+        try {
+            CompletableFuture<DeviceTwinName> twinNamesTask = this.iotHubClient.GetDeviceTwinNamesAsync().toCompletableFuture();
+            CompletableFuture<HashSet<String>> reportedNamesTask = this.simulationClient.GetDevicePropertyNamesAsync().toCompletableFuture();
+            CompletableFuture.allOf(twinNamesTask, reportedNamesTask).get();
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java]: Considering that the code runs in the background, there's no need to run these in parallel. We should split this code in 2 blocks, one for twin and one for property names, in separate try catch blocks, with specific error messages in the logs.

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java [hunk 23]
```diff
@@ -129,66 +116,55 @@ public void run() {
...
+            ValueApiModel response = this.storageClient.updateAsync(CacheCollectionId, CacheKey,
+                    Json.stringify(Json.toJson(new CacheValue(null, null, false))), null).
+                    toCompletableFuture().get();
+            etag = response.getETag();
+        }
+        String value = Json.stringify(Json.toJson(new CacheValue(twinNames.getTags(), reportedNames, false)));
+        try {
+            return this.storageClient.updateAsync(CacheCollectionId, CacheKey, value, etag);
+        } catch (ConflictingResourceException e) {
+            log.warn("rebuild Conflicted ");
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java]: include more details

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/ICache.java [hunk 24]
```diff
@@ -16,9 +16,9 @@
 
     CompletionStage<CacheValue> SetCacheAsync(CacheValue cache) throws BaseException, ExecutionException, InterruptedException;
 
-    CompletionStage RebuildCacheAsync(boolean force) throws BaseException, ExecutionException, InterruptedException, URISyntaxException;
+    CompletionStage RebuildCacheAsync(boolean force) throws Exception;
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/ICache.java]: we should specify explicitly all the checked exceptions, otherwise it becomes impossible to handle them appropriately upstream

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/ICache.java [hunk 25]
```diff
@@ -16,9 +16,9 @@
...
 
-    default public CompletionStage RebuildCacheAsync() throws InterruptedException, ExecutionException, BaseException, URISyntaxException {
+    default public CompletionStage RebuildCacheAsync() throws Exception {
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/ICache.java]: we should specify explicitly all the checked exceptions, otherwise it becomes impossible to handle them appropriately upstream

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/RecurringTasks.java [hunk 26]
```diff
@@ -0,0 +1,124 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/RecurringTasks.java]: code looks good 👍

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java [hunk 27]
```diff
@@ -34,7 +35,7 @@
     private final String SeedCollectionId = "solution-settings";
     private final String MutexKey = "seedMutex";
     private final String CompletedFlagKey = "seedCompleted";
-    private int mutexTimeout = 60 * 5;
+    private int mutexTimeout = 60 * 5;// seconds
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: rename the var to make the code readable also later, so we don't need comments

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/external/ValueApiModel.java (hunk 16 above)
Comment (dluc) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/external/ValueApiModel.java]: I'm sad, this typo is spreading everywhere, and we're not making anything to improve the situation

File: app/com/microsoft/azure/iotsolutions/uiconfig/webservice/v1/controllers/DeviceGroupFiltersController.java [hunk 28]
```diff
@@ -42,7 +42,7 @@ public DeviceGroupFiltersController(ICache cache) {
     }
 
     @With(DepressedFilter.class)
-    public CompletionStage<Result> rebuildAsync() throws BaseException, InterruptedException, ExecutionException, URISyntaxException {
+    public CompletionStage<Result> rebuildAsync() throws Exception {
```
Comment (dluc) [inline app/com/microsoft/azure/iotsolutions/uiconfig/webservice/v1/controllers/DeviceGroupFiltersController.java]: ditto for explicit exceptions types

File: app/Module.java (hunk 19 above)
Comment (v-limliu) [reply app/Module.java]: // global setting is not recommend for application_onStart event, PLS refer here for details :https://www.playframework.com/documentation/2.6.x/GlobalSettings

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/external/ValueApiModel.java (hunk 16 above)
Comment (zhang-hua) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/external/ValueApiModel.java]: we need to fix the annoying ETag naming issue once for all projects. It is a typical issue to coordinating so may projects simultaneously.In the future, we might consider to establish a separate API spec repo and keep to update it so that everyone can align with it timely.

File: app/Module.java (hunk 20 above)
Comment (v-limliu) [reply app/Module.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java (hunk 27 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Seed.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java (hunk 23 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java (hunk 22 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java (hunk 21 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/Cache.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/ICache.java (hunk 24 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/ICache.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/services/ICache.java (hunk 25 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/services/ICache.java]: fixed

File: app/com/microsoft/azure/iotsolutions/uiconfig/webservice/v1/controllers/DeviceGroupFiltersController.java (hunk 28 above)
Comment (v-limliu) [reply app/com/microsoft/azure/iotsolutions/uiconfig/webservice/v1/controllers/DeviceGroupFiltersController.java]: fixed
